pallets/stateful-storage/src/stateful_child_tree.rs

```diff
@@ -1,37 +1,238 @@
-use codec::{Decode, Encode};
+use core::marker::PhantomData;
+
+use codec::{Decode, Encode, MaxEncodedLen};
 use common_primitives::msa::MessageSourceId;
 use frame_support::{
 	storage::{child, child::ChildInfo, ChildTriePrefixIterator},
-	Identity, StorageHasher,
+	Blake2_128, Blake2_256, Hashable, StorageHasher, Twox128, Twox256,
 };
-use sp_core::hexdisplay::AsBytesRef;
-use sp_std::prelude::*;
+use sp_core::{ConstU8, Get};
+use sp_std::{fmt::Debug, prelude::*};
 
-pub type StatefulPageKeyPart = Vec<u8>;
+pub trait MultipartKeyStorageHasher: StorageHasher {
+	type HashSize: Get<u8>;
+	fn reverse(x: &[u8], num_key_parts: u8) -> &[u8] {
+		let hash_len: usize = (Self::HashSize::get() * num_key_parts).into();
+		if x.len() < hash_len {
+			log::error!("Invalid reverse: hash length too short");
+			return &[]
+		}
+		&x[hash_len..]
+	}
+}
+/// 128-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit Blake2 hash.
+impl MultipartKeyStorageHasher for Blake2_256 {
+	type HashSize = ConstU8<32>;
+}
+/// 128-bit XX hash.
+impl MultipartKeyStorageHasher for Twox128 {
+	type HashSize = ConstU8<16>;
+}
+/// 256-bit XX hash.
+impl MultipartKeyStorageHasher for Twox256 {
+	type HashSize = ConstU8<32>;
+}
 
-/// Paginated Stateful data access utility
-pub struct StatefulChildTree;
-impl StatefulChildTree {
-	pub fn concat_keys(keys: &[StatefulPageKeyPart]) -> Vec<u8> {
-		let mut key = Vec::<u8>::new();
-		for k in keys {
-			let mut k2 = k.to_owned();
-			key.append(&mut k2);
+pub trait IsStatefulKey<H: MultipartKeyStorageHasher>:
```

Comment:
This trait does more than check if something is a stateful key. I don't have any suggestions for naming at the moment.

Reply:
Hmm, good point. It kind of grew. Maybe `MultipartKey` is a bettername?

Reply:
Renamed to `MultipartKey`